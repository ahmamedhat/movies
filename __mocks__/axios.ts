module.exports = {
    get: jest.fn().mockResolvedValue(),
    post: jest.fn().mockResolvedValue(),
    create: jest.fn(function () {
        return this;
    }),
    interceptors: {
        request: { use: jest.fn(), eject: jest.fn() },
        response: { use: jest.fn(), eject: jest.fn() }
    }
};