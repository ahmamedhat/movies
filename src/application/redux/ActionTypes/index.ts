export const SET_CURRENT_USER = 'SET_CURRENT_USER'
export const SET_CURRENT_LANGUAGE = 'SET_CURRENT_LANGUAGE'
