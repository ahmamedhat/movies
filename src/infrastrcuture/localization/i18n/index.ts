import I18n from "react-native-i18n"
const en = require("../../../translations/en")
const ar = require("../../../translations/ar")
I18n.fallbacks = true;

I18n.translations = {
    en: en,
    ar: ar,
};

export default I18n