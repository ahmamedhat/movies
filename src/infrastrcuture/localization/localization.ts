import I18n from "../../infrastrcuture/localization/i18n"

export function localizedString(key: string): string {
    return I18n.t(key)
}