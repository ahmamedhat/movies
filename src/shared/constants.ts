export class AsyncStorageKeys {
    static USER_KEY = '@user_Key'
    static LANGUAGE_KEY = '@language_Key'
}

export enum LanguageCode {
    EN = 'en',
    AR = 'ar'
}

export enum Endpoints {
    MOVIES = 'movie'
}


export enum TestIds {
    TEST_ID_LOGIN_EMAIL_INPUT = 'loginEmailInput',
    TEST_ID_LOGIN_PASSWORD_INPUT = 'loginPasswordInput',
    TEST_ID_LOGIN_SUBMIT_BUTTON = 'loginSubmitButton',
}

export enum LocalizationKeys {
    EMAIL = 'email',
    PASSWORD = 'password',
    SUBMIT = 'submit',
    GENERAL_SERVER_ERROR = 'generalError',
    INVALID_EMAIL = 'invalidEmail',
}