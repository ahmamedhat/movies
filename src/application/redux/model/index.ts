import { User } from "../../../domain/user";
import { LanguageCode } from "../../../shared/constants";

export interface UserReducer {
    currentUser: User
}
export interface LanguageReducer {
    currentLanguage: LanguageCode
}

export interface RootState {
    user: UserReducer
    language: LanguageReducer
}