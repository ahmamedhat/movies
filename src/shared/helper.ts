import { AxiosError } from "axios";
import { Alert } from "react-native";
import { ServerError } from "../domain/ServerError";
import { localizedString } from "../infrastrcuture/localization/localization";
import { LocalizationKeys } from "./constants";

export function parseServerError(error: AxiosError<ServerError>) {
    return error?.response?.data?.errors?.message ?? localizedString(LocalizationKeys.GENERAL_SERVER_ERROR)
}

export function notifiyUser(message: string) {
    Alert.alert(message)
}