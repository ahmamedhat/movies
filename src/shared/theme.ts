import { LanguageCode } from './constants'
export class Font {
    static FONT_1 = ""
    static FONT_2 = ""

    static getFontForLang = (lang: LanguageCode) => {
        return lang === LanguageCode.EN ? Font.FONT_1 : Font.FONT_2
    }
}

export class Color {
    static PRIMARY_COLOR = "red"
    static SECOUNDARY_COLOR = "blue"
}