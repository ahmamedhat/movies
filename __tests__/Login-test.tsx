/**
 * @format
 */

import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import LoginScreen from '../src/presentation/screens/Login/Login';
import { render, fireEvent } from "@testing-library/react-native"
import { AsyncStorageKeys, Endpoints, LocalizationKeys, TestIds } from '../src/shared/constants';
import { Alert } from 'react-native';
import { localizedString } from '../src/infrastrcuture/localization/localization';
import AsyncStorage from '@react-native-async-storage/async-storage';
import axios from 'axios'
import * as reactRedux from 'react-redux'
import { SET_CURRENT_USER } from '../src/application/redux/ActionTypes';
import { mockDispatch } from '../jest.mock';

beforeEach(() => {
  jest.clearAllMocks()
})

describe('Login Screen', () => {
  it('renders correctly', () => {
    const screen = renderer.create(<LoginScreen />).toJSON()
    expect(screen).toMatchSnapshot()
  });
  it('show Invalid email', () => {
    const email = 'email'
    const { getByTestId } = render(<LoginScreen />)
    const emailInput = getByTestId(TestIds.TEST_ID_LOGIN_EMAIL_INPUT)
    const submitButtom = getByTestId(TestIds.TEST_ID_LOGIN_SUBMIT_BUTTON)
    fireEvent.changeText(emailInput, email)
    fireEvent.press(submitButtom)
    expect(Alert.alert).toHaveBeenCalledTimes(1)
    expect(Alert.alert).toHaveBeenCalledWith(localizedString(LocalizationKeys.INVALID_EMAIL))
  });
  it('handle valid user', async () => {
    const email = 'email@address.com'
    const pwd = 'secret'
    const { getByTestId } = render(<LoginScreen />)
    const emailInput = getByTestId(TestIds.TEST_ID_LOGIN_EMAIL_INPUT)
    const pwdInput = getByTestId(TestIds.TEST_ID_LOGIN_PASSWORD_INPUT)
    const submitButtom = getByTestId(TestIds.TEST_ID_LOGIN_SUBMIT_BUTTON)
    fireEvent.changeText(emailInput, email)
    fireEvent.changeText(pwdInput, pwd)
    let response = {
      data: { email: email, password: pwd }
    }
    axios.post.mockResolvedValueOnce(response)
    await fireEvent.press(submitButtom)
    expect(axios.post).toHaveBeenCalledWith(Endpoints.LOGIN, { email: email, password: pwd })
    expect(AsyncStorage.setItem).toHaveBeenCalledWith(AsyncStorageKeys.USER_KEY, JSON.stringify(response.data))
    expect(mockDispatch).toHaveBeenCalledWith({ type: SET_CURRENT_USER, payload: response.data })
  });
  it('handle invalid user', async () => {
    const email = 'email@address.com'
    const pwd = 'aaa'
    const { getByTestId } = render(<LoginScreen />)
    const emailInput = getByTestId(TestIds.TEST_ID_LOGIN_EMAIL_INPUT)
    const pwdInput = getByTestId(TestIds.TEST_ID_LOGIN_PASSWORD_INPUT)
    const submitButtom = getByTestId(TestIds.TEST_ID_LOGIN_SUBMIT_BUTTON)
    fireEvent.changeText(emailInput, email)
    fireEvent.changeText(pwdInput, pwd)
    let response = {
      status: 401,
      response: { data: { "errors": { "message": "Invalid Credentials", "code": 101 } } }
    }
    axios.post.mockRejectedValueOnce(response)
    await fireEvent.press(submitButtom)
    expect(axios.post).toHaveBeenCalledWith(Endpoints.LOGIN, { email: email, password: pwd })
    expect(AsyncStorage.setItem).not.toBeCalled()
    expect(Alert.alert).toHaveBeenCalledWith('Invalid Credentials')
  });
  it('handle unxepected error response from api', async () => {
    const email = 'email@address.com'
    const pwd = 'aaa'
    const { getByTestId } = render(<LoginScreen />)
    const emailInput = getByTestId(TestIds.TEST_ID_LOGIN_EMAIL_INPUT)
    const pwdInput = getByTestId(TestIds.TEST_ID_LOGIN_PASSWORD_INPUT)
    const submitButtom = getByTestId(TestIds.TEST_ID_LOGIN_SUBMIT_BUTTON)
    fireEvent.changeText(emailInput, email)
    fireEvent.changeText(pwdInput, pwd)
    let response = {
      status: 500,
      response: ''
    }
    axios.post.mockRejectedValueOnce(response)
    await fireEvent.press(submitButtom)
    expect(axios.post).toHaveBeenCalledWith(Endpoints.LOGIN, { email: email, password: pwd })
    expect(AsyncStorage.setItem).not.toBeCalled()
    expect(Alert.alert).toHaveBeenCalledWith(localizedString(LocalizationKeys.GENERAL_SERVER_ERROR))
  });
  it('handle unexpected success response from api', async () => {
    const email = 'email@address.com'
    const pwd = 'secret'
    const { getByTestId } = render(<LoginScreen />)
    const emailInput = getByTestId(TestIds.TEST_ID_LOGIN_EMAIL_INPUT)
    const pwdInput = getByTestId(TestIds.TEST_ID_LOGIN_PASSWORD_INPUT)
    const submitButtom = getByTestId(TestIds.TEST_ID_LOGIN_SUBMIT_BUTTON)
    fireEvent.changeText(emailInput, email)
    fireEvent.changeText(pwdInput, pwd)
    let response = {
      data: ''
    }
    axios.post.mockResolvedValueOnce(response)
    await fireEvent.press(submitButtom)
    expect(axios.post).toHaveBeenCalledWith(Endpoints.LOGIN, { email: email, password: pwd })
    expect(AsyncStorage.setItem).not.toBeCalled()
    expect(Alert.alert).toHaveBeenCalledWith(localizedString(LocalizationKeys.GENERAL_SERVER_ERROR))
  });
})
