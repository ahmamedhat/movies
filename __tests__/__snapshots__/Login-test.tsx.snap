// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCorrect={false}
    keyboardType="email-address"
    onChangeText={[Function]}
    placeholder="Email"
    rejectResponderTermination={true}
    style={
      Object {
        "borderColor": "blue",
        "borderWidth": 1,
        "height": 40,
        "marginHorizontal": 16,
        "padding": 4,
      }
    }
    testID="loginEmailInput"
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Password"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Object {
        "borderColor": "blue",
        "borderWidth": 1,
        "height": 40,
        "marginHorizontal": 16,
        "marginTop": 16,
        "padding": 4,
      }
    }
    testID="loginPasswordInput"
    underlineColorAndroid="transparent"
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="loginSubmitButton"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        submit
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
